source/s3_meta_request.c

```diff
@@ -842,20 +832,269 @@ static void s_s3_meta_request_send_request_finish_default(
+        error_type = AWS_RETRY_ERROR_TYPE_SERVER_ERROR;
+    } else if (error_code == AWS_ERROR_S3_SLOW_DOWN) {
+        error_type = AWS_RETRY_ERROR_TYPE_THROTTLING;
+    } else {
+        error_type = AWS_RETRY_ERROR_TYPE_TRANSIENT;
+    }
+
+    request->retry_token = token;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_S3_META_REQUEST,
+        "id=%p Meta request scheduling retry of request %p with token %p",
+        (void *)meta_request,
+        (void *)request,
+        (void *)token);
+
+    /* Ask the retry strategy to schedule a retry of the request. */
+    if (aws_retry_strategy_schedule_retry(request->retry_token, error_type, s_s3_meta_request_retry_ready, request)) {
+        error_code = aws_last_error();
+
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p Meta request could not retry request %p with token %p due to error %d (%s)",
+            (void *)meta_request,
+            (void *)request,
+            (void *)token,
+            error_code,
+            aws_error_str(error_code));
+
+        goto error_ref_clean_up;
+    }
+
+    return;
+
+error_ref_clean_up:
+
+    aws_s3_request_release(request);
+    aws_s3_meta_request_release(meta_request);
+
+    aws_s3_meta_request_finish(meta_request, NULL, 0, aws_last_error());
+}
+
+static void s_s3_meta_request_acquire_retry_token(
+    struct aws_retry_strategy *retry_strategy,
+    struct aws_s3_request *request) {
+    AWS_PRECONDITION(retry_strategy);
+    AWS_PRECONDITION(request);
+
+    struct aws_s3_meta_request *meta_request = request->meta_request;
+    AWS_PRECONDITION(meta_request);
+
+    struct aws_http_message *message = meta_request->initial_request_message;
+    AWS_PRECONDITION(message);
+
+    struct aws_byte_cursor host_header_name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Host");
+
+    struct aws_http_headers *message_headers = aws_http_message_get_headers(message);
+    AWS_ASSERT(message_headers);
+
+    struct aws_byte_cursor host_header_value;
+
+    bool host_header_exists =
+        aws_http_headers_get(message_headers, host_header_name, &host_header_value) == AWS_ERROR_SUCCESS;
+    AWS_ASSERT(host_header_exists);
+    (void)host_header_exists;
+
+    /* Try to acquire a token so that we can schedule a retry. */
+    if (aws_retry_strategy_acquire_retry_token(
+            retry_strategy, &host_header_value, s_s3_meta_request_queue_retry_with_token, request, 0)) {
+
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p Meta request could not acquire retry token for request %p due to error %d (%s)",
+            (void *)request->meta_request,
+            (void *)request,
+            aws_last_error(),
+            aws_error_str(aws_last_error()));
+
+        goto error_ref_clean_up;
+    }
+
+    return;
+
+error_ref_clean_up:
+
+    aws_s3_request_release(request);
+    aws_s3_meta_request_finish(meta_request, NULL, 0, aws_last_error());
```

Comment:
If aws_s3_meta_request_finish judges success/failure via the error_code, I would strongly suggest never passing aws_last_error() in but aws_last_error_or_unknown() to guarantee the failure path is taken even when your downstream calls forget to raise_error on a failure.

Reply:
This is coming in the already-approved signing PR.  This PR just needs to go in first.